contracts/verifying/VerifyingSingletonPaymasterV2.sol

```diff
@@ -1,9 +1,363 @@
-// SPDX-License-Identifier: MIT
-pragma solidity ^0.8.17;
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.17;
 
-// todo Add pricing
-// Instead of paymaster helpers write parsePaymasterData methods
+/* solhint-disable reason-string */
+/* solhint-disable no-inline-assembly */
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {UserOperation, UserOperationLib} from "@account-abstraction/contracts/interfaces/UserOperation.sol";
+import "../BasePaymaster.sol";
+import {VerifyingPaymasterErrors} from "../common/Errors.sol";
+import {IVerifyingSingletonPaymaster} from "../interfaces/paymasters/IVerifyingSingletonPaymaster.sol";
 
-contract VerifyingSingletonPaymasterV2 {
+contract VerifyingSingletonPaymasterV2 is
```

Comment:
can we not extend from earlier paymaster and just override methods with the logic changes? That way it'll be very clear what are the changes from last version of paymaster

Reply:
no, as discussed there are other improvements in this paymaster which removes decoding using helpers and adding parsePaymasterAndData method.

Reply:
**diff view**

https://www.diffchecker.com/5HuMm7zG/

**Overview**

- removed library PaymasterHelpers. as we don't need helpers for decoding and encoding context or paymasterAndData [ We follow this in token paymaster as well + refer from infinitism Samples ]

- moved all the events + interface methods (which are not there in BasePaymaster abstract class already and are specific to singleton aspect) in IVerifyingSingletonPaymaster.sol

-  paymasterAndData 

    was: concat of [paymasterAddress(address), abi.encode(paymasterId, validUntil, validAfter, signature)]
    
    now: concat of [paymasterAddress(address), abi.encode(paymasterId, validUntil, validAfter, priceMarkup), signature]
   
- parsePaymasterAndData helper added to split paymasterAndData as per above logic. Two constants are added for offsets (again pattern used in BTPM and infinitism samples)

- Added new state variable fixedPriceMarkup. also taken in the constructor (can be optional). also setter for this

- getHash will also take another attribute priceMarkup which is dynamic markup and MUST be signed by paymaster verifying signer (using paymaster service)

- helpers to get min and max for different int types

- validation stage considers markup as well to be able to charge premium. when we check paymasterId balances. dynamic markup is calculated based on provided vs fixed

- context includes price markup

- postOp applies effective dynamic price markup (check validation notes above) to the actual gas cost and deducts deposit. plus it adjust the deposit of the owner = paymasterId

- new event FeeCollected is emitted